File: backend/app/schemas/devoluciones.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class DevolucionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    serial: str
    nombre: str | None = None
    telefono: str | None = None
    direccion: str | None = None
    localidad: str | None = None
    estado: str
    fecha_carga: datetime
    fecha_actualizacion: datetime


class DevolucionCreate(BaseModel):
    serial: str = Field(min_length=1, max_length=50)
    nombre: str | None = Field(default=None, max_length=255)
    telefono: str | None = Field(default=None, max_length=20)
    direccion: str | None = None
    localidad: str | None = Field(default=None, max_length=100)
    estado: str = Field(default="transito", min_length=1, max_length=30)


class DevolucionEstadoUpdate(BaseModel):
    estado: str = Field(min_length=1, max_length=30)


class CargaMasivaDevolucionesResult(BaseModel):
    total_filas: int
    nuevas: int
    actualizadas: int
    errores: list[str]


class DevolucionDocumentoItem(BaseModel):
    serial: str
    nombre: str | None = None
    direccion: str | None = None
    localidad: str | None = None


class DevolucionDocumentoRequest(BaseModel):
    items: list[DevolucionDocumentoItem] = Field(min_length=1)


class VerificarSerialesRequest(BaseModel):
    seriales: list[str] = Field(min_length=1, max_length=500)


class SerialVerificado(BaseModel):
    serial: str
    clasificacion: Literal["entrega", "devolucion", "ninguna"]
    fuente: Literal["seriales_gestion", "devoluciones"] | None = None
    estado_detalle: str | None = None
    cliente: str | None = None
    planilla: str | None = None
    cod_men: str | None = None
    fecha: date | None = None


class VerificarSerialesResult(BaseModel):
    items: list[SerialVerificado]
    total_devoluciones: int
    total_entregas: int
    total_ninguna: int

File: frontend/src/api/devoluciones.ts
import api from "./client";

export interface Devolucion {
  id: number;
  serial: string;
  nombre: string | null;
  telefono: string | null;
  direccion: string | null;
  localidad: string | null;
  estado: string;
  fecha_carga: string;
  fecha_actualizacion: string;
  fecha_escaneo: string | null;
}

export interface DevolucionEscaneo extends Devolucion {
  ya_escaneado: boolean;
  escaneado_previamente_en: string | null;
}

export interface CargaMasivaDevolucionesResult {
  total_filas: number;
  nuevas: number;
  actualizadas: number;
  errores: string[];
}

export interface SerialVerificado {
  serial: string;
  clasificacion: "entrega" | "devolucion" | "ninguna";
  fuente: "seriales_gestion" | "devoluciones" | null;
  estado_detalle: string | null;
  cliente: string | null;
  planilla: string | null;
  cod_men: string | null;
  fecha: string | null;
}

export interface VerificarSerialesResult {
  items: SerialVerificado[];
  total_devoluciones: number;
  total_entregas: number;
  total_ninguna: number;
}

export const devolucionesApi = {
  list: (params?: { estado?: string; q?: string; limit?: number; offset?: number }) =>
    api.get<Devolucion[]>("/devoluciones/", { params }),

  create: (data: {
    serial: string;
    nombre?: string;
    telefono?: string;
    direccion?: string;
    localidad?: string;
    estado?: string;
  }) => api.post<Devolucion>("/devoluciones/", data),

  cargaMasiva: (file: File) => {
    const form = new FormData();
    form.append("file", file);
    return api.post<CargaMasivaDevolucionesResult>("/devoluciones/carga-masiva", form, {
      headers: { "Content-Type": "multipart/form-data" },
    });
  },

  updateEstado: (id: number, estado: string) =>
    api.patch<Devolucion>(`/devoluciones/${id}`, { estado }),

  escanearSerial: (serial: string) =>
    api.patch<DevolucionEscaneo>(`/devoluciones/serial/${encodeURIComponent(serial)}`, {
      estado: "devolucion",
    }),

  escaneadosDia: (fecha?: string) =>
    api.get<Devolucion[]>("/devoluciones/escaneados-dia", {
      params: fecha ? { fecha } : undefined,
    }),

  generarDocumento: (
    items: { serial: string; nombre: string | null; direccion: string | null; localidad: string | null }[]
  ) => api.post("/devoluciones/documento", { items }, { responseType: "blob" }),

  reporteDia: (fecha: string) =>
    api.get("/devoluciones/reporte-dia", { params: { fecha }, responseType: "blob" }),

  reporteDiaWord: (fecha: string) =>
    api.get("/devoluciones/reporte-dia/word", { params: { fecha }, responseType: "blob" }),

  reporteDiaExcel: (fecha: string) =>
    api.get("/devoluciones/reporte-dia/excel", { params: { fecha }, responseType: "blob" }),

  verificarSeriales: (seriales: string[]) =>
    api.post<VerificarSerialesResult>("/devoluciones/verificar-seriales", { seriales }),

  verificarSerialesExcel: (file: File) => {
    const form = new FormData();
    form.append("file", file);
    return api.post("/devoluciones/verificar-seriales-excel", form, {
      headers: { "Content-Type": "multipart/form-data" },
      responseType: "blob",
    });
  },
};
